Create a PNG image output object

Hysrok opened the pull request:
Adds the ability to create PNG files from the data provided by input files.
Can specify desired resolution.
Default output is grayscale, but can be put into color (current color scheme 'low -> high' is 'blue->red->yellow->white')
Can specify to save one specific step PNG (default is to save all steps with filename plus numerical identifier).

Closes #12846

Commit:
Created a class to create png files based on the tests that are run.

Closes #12846

permcody commented on test/tests/kernels/simple_transient_diffusion/simple_transient_diffusion.i:
```diff
@@ -46,6 +56,6 @@
-[]
+#[Outputs]
+#  exodus = true
+#[]
```
Oops!

permcody commented on test/tests/kernels/simple_transient_diffusion/simple_transient_diffusion.i:
```diff
@@ -1,3 +1,13 @@
+[Outputs]
+  [png]
+    type = PNGObject
+    PNGFile = "PNGTEST"
```
Let's not update this test. Create a new test (or two or three) in the test/tests/outputs/png/ folder (or something similar)

permcody commented on test/tests/kernels/adv_diff_reaction/adv_diff_reaction_test.i:
```diff
@@ -113,6 +113,12 @@
+    type = PNGObject
+    PNGFile = "PNGTEST"
+    resolution = 50
+    inColor = false
```
Not this test either but feel free to copy this test into the outputs directory to create a new test (see comment below)

permcody commented on test/tests/bcs/periodic/wedge.i:
```diff
@@ -70,8 +70,13 @@
 []
 
 [Outputs]
+  [png]
```
Not this test either

permcody commented on test/Makefile:
```diff
@@ -16,6 +16,9 @@ ADDITIONAL_CPPFLAGS += -Wall -Wextra
 include $(FRAMEWORK_DIR)/moose.mk
 
+ADDITIONAL_LIBS := "-L/opt/moose/libpng-1.6.25/lib -lz -lpng16"
+ADDITIONAL_INCLUDES := "-I/opt/moose/libpng-1.6.25/include/libpng16"
```
This is the right idea but we can't hardcode these paths. I can work on this for you.

permcody replied:
Make sure you delete these lines!

permcody commented on framework/src/outputs/PNGTest.C:
```diff
@@ -0,0 +1,279 @@
+//   row = (png_bytep) malloc(width * sizeof(png_byte));
+  // int grayscale[((int)(resolution*resolution))];
+  // for (int i = 0; i < resolution*resolution; i++)
+  //   grayscale[i] = rand() % 256;
```
Go ahead and clean up your comments. If you want them around in history feel free to add another commit that removes them that you don't squash. That way they will be in the intermediate commit.

permcody commented on framework/src/outputs/PNGTest.C:
```diff
@@ -0,0 +1,279 @@
+
+  pngp = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+  if(!pngp) {
+    std::cout << "Failed to make pointer string.";
```
mooseError()

permcody commented on framework/src/outputs/PNGTest.C:
```diff
@@ -0,0 +1,279 @@
+
+  infop = png_create_info_struct(pngp);
+  if(!infop) {
+    std::cout << "Failed to make info pointer.";
```
mooseError()

permcody commented on framework/src/outputs/PNGTest.C:
```diff
@@ -0,0 +1,279 @@
+  }
+
+  else
+    std::cout << "Nothing selected.";
```
These stream outputs need to be dropped - your object should probably be silent.

permcody commented on framework/src/outputs/PNGTest.C:
```diff
@@ -0,0 +1,279 @@
+ 	int color = (int)(selection * 767);
+  // If this isn't true, something off with our scaling function,
+  // because everything should be between 1 and 0. (Or math is broken)
+ 	assert(color > 0 && color < 767);
```
What is the significance of 767? Can you create a named constant?

permcody replied:
Also, use `mooseAssert(condition, error message)` instead of assert.

permcody reviewed (changes requested):
I'm not going to repeat everything in every case. Go ahead and go over this PR and make sure you address each of my comments everywhere where it applies. Let me know if anything doesn't make sense. Nice work!

dschwen commented:
How about calling this `PNGOutput`? "Object" is a little strange.

dschwen commented on framework/include/outputs/PNGObject.h:
```diff
@@ -0,0 +1,67 @@
+  PNGObject(const InputParameters & parameters);
+
+  // Method for assigning color values to the PNG
+  void setRGB(png_byte *rgb, float selection);
```
How is this working? You are not including libpng stuff here but use this type in a public interface. Does it have to be public in the first place?

permcody replied:
Probably doesn't need to be public. This PR has a ways to go. I'm helping Sam get the right things into the Makefile and then we'll continue to work on the interface.

Hysrok replied:
Yeah, I don't see a reason any of it should be public except the constructor.  Public was just easier to start.  Right now I switched everything except the constructor to protected.  Not sure if any of it should be private yet.  I haven't really thought about what other classes would need from this class, if they'd ever need anything at all.

permcody committed:
Added newest changes from the Makefile branch.

- Added include path to the header to make sure we get the right version 
of png.h.

permcody committed:
Adding in optional linking and compiler defines for libPNG

refs #12846

Commit:
General clean up and fixes to PNGOutput

- Fixed some formatting errors.
- Renamed the classes.
- Created tests that don't change the original tests.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/359354/) on 1ca0747 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fe5780e2a7533ab32b14516ba99ae351771ff823`

dschwen commented:
@Hysrok you might want to look into configuring your editor to automatically run clang-format on save...  
Here is what I use: https://www.mooseframework.org/help/development/Atom_Editor.html

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+
+  // Create the filename based on base and the test step number.
+  std::string PNGFile2 = PNGFile;
+  PNGFile2 += std::to_string((int)testStep);
```
Good - however, we generally like padding so that sorting works better. We typically do `foo_001.png foo_002.png, etc`. You can do this with `std::setw` and `std::fill`.

Hysrok replied:
This was confusing me a bit, but I think I can get it once I find the right way to use _file_base.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/359528/) on 76315c2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c3f0d07dfc87f2943c756451e920c12865de7bc5`

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  makePNG();
+}
+
+void PNGOutput::makePNG() {
```
Make sure this routine only runs on rank 0. `_communicator.processor_id()` maybe even just `processor_id()`

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  PNGFile2 += ".png";
+
+
+  FILE * fp = nullptr;
```
This type comes from a Unix header somewhere. It's a C type. Are we locked into this type for the png library? If not there's a C++ type for files (`ofstream`).

Hysrok replied:
I've been looking into/working on this, and it looks like it is possible to get out of using the FILE*, but I've been having a lot of difficulty actually getting anything else to work.  So should we just want to go with what we have for now and try changing it later, or if I should focus on this more right now?

Hysrok replied:
So I did some more looking into this and working to try to get it to work.  I read something that said that it has to have a FILE* in order to work, but I understood libpng manual to be saying that there was a way around it.  Specifically if you tell it that you have a function that you want it to use in order to write the image.  I haven't been able to get it to work at all yet.

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+    if(testStep != testStepToPNG)
+      return;
+
+  fp = fopen(PNGFile2.c_str(), "wb");
```
We have a method for checking writability in MooseUtils.h. Use that instead.

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  if(!fp)
+  {
+    mooseError("Failed to open the file for the PNG output.");
+    return;
```
No need to do anything after a `mooseError()`. It will halt the program (no early return).

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  png_write_info(pngp, infop);
+
+  // Allocate resources.
+  row = new png_byte[(width * 3)+1];
```
We are really moving away from raw memory management in MOOSE. This is all local to the routine so it's not a big deal but there are a couple other arguably better ways to handle this. 

1) Stash your raw array allocation in a std::unique_ptr() - there are methods for this.
2) Use a std::vector and the std::vector::data method to get back a pointer to the raw vector.

Hysrok replied:
Ok, so I couldn't quite get it to work with the uniqu_ptr, but I think I got it to work with the vector.

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  // Allocate resources.
+  row = new png_byte[(width * 3)+1];
+
+  Point pt(1,1,0);
```
This line should probably have a comment. Why 1,1,0?

Hysrok replied:
No real reason, except to make sure that the z value is 0.  The other part was just from earlier testing, but now they're getting set to their proper starting values before the pt is used.

Commit:
Applied clang-format

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,237 @@
+
+#define nls _problem_ptr->getNonlinearSystem()
+#define scaledValue ((dv(0) + shiftValue) / scalingMax)
+#define reverseScale ((outBoundsControlColor * scalingMax) - shiftValue)
```
Let's avoid the macros. None of them are necessary and they are context sensitive. Go ahead and create new variables or functions instead of using these.

Hysrok replied:
So I changed the scaledValue and reverseScale into inline functions, and I replaced nls macro symbol with the macro value everywhere it occurred. I kept running into problems with the other things I tried to do to create a variable that would end up giving the same value as the macro.

permcody commented on test/tests/outputs/png/adv_diff_reaction_test.i:
```diff
@@ -0,0 +1,124 @@
+[Mesh]
```
Great! Now you need to add the "tests" file so that these actually get run. I'm going to suggest a "CheckFiles" test but we might be able to do better. 

@aeslaughter - Can we do a png difference or is that only available within the python testing infrastructure?

permcody commented on test/tests/outputs/png/adv_diff_reaction_test.i:
```diff
@@ -0,0 +1,124 @@
+    resolution = 50
+    inColor = false
+  []
+  file_base = out
```
Let's drop the `file_base` in each of these. That's a very old pattern that we discourage now. We typically leave this off for testing unless we have a good reason. By default you'll get outputs based on the input file name instead, which is more intuitive when you are trying to match which outputs come from which inputs. Also go ahead and drop the `exodus = true`. Your tests should just be testing the PNG capability.

permcody commented on test/tests/outputs/png/wedge.i:
```diff
@@ -0,0 +1,82 @@
+    resolution = 250
+    inColor = false
+  []
+  execute_on = 'timestep_end'
```
Drop this line

permcody commented on test/tests/outputs/png/adv_diff_reaction_test.i:
```diff
@@ -0,0 +1,124 @@
+[Mesh]
+  dim              = 2
+  file             = Mesh24.e
```
You'll need to copy this mesh file over into this directory to run.

permcody reviewed (changes requested):
OK - this is my thorough review

dschwen commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+  InputParameters params = validParams<Output>();
+  params.addParam<Real>("resolution", 2000, "The resolution of the image.");
+  params.addParam<std::string>("PNGFile", "Adam", "Root filename of the PNG to be created.");
+  params.addParam<Real>("testStepToPNG", -1, "PNG to save.");
```
We have barely any camelCase parameter names. Our de facto standard is using underscores, so this should be `test_step_to_png` (if you need this parameter at all, but same applies for the other parameters).

dschwen commented on framework/include/outputs/PNGOutput.h:
```diff
@@ -0,0 +1,73 @@
+  Real resolution;
+
+  // Name of file to which the PNG will be written.
+  std::string PNGFile;
```
As Cody said: prefix with an underscore. **But** make it lowercase! Symbol names in any scope that begin with an underscore and are followed by a capital letter are reserved for use as implementation macros. We should really get away from adding more of those names.

dschwen commented on test/tests/outputs/png/wedge.i:
```diff
@@ -0,0 +1,82 @@
+    inColor = false
+  []
+  execute_on = 'timestep_end'
+  file_base = out_wedge
```
could we just agree not to use `file_base` in tests for no reason. It is so much simpler if the output files follow the automatic moose convention (and the test names as well for that matter). Otherwise you always have to consult the input file to know which outout belongs to which test.

permcody replied:
I do agree - I actively discourage it. Going back and fixing up all the existing tests will take time.

Hysrok commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+    rgb[0] = 255;
+    rgb[1] = 255;
+    rgb[2] = magnitude;
+  }
```
It's based on how some others were using color with libpng in some examples I was looking through.  I think having several color schemes would be pretty nice.

Hysrok replied:
As commented below by dschwen, Spectrum could possibly be really nice for this?

Hysrok replied:
I added a few more color schemes as well as adding a MooseEnum to be able to switch between color schemes.

Hysrok commented on framework/include/outputs/PNGOutput.h:
```diff
@@ -0,0 +1,73 @@
+  bool inColor;
+
+  // Way to track the number test step number and the associated png, if created.
+  Real testStep = 0;
```
Yeah, all I was using this for was to differentiate between images by giving a separate value to each string name.  As you mentioned this won't work right in intervals and repeated steps.

Hysrok commented on framework/include/outputs/PNGOutput.h:
```diff
@@ -0,0 +1,73 @@
+InputParameters validParams<PNGOutput>();
+
+// pnglib include
+#include <png.h>
```
The setRGB function uses a "png_byte".  It's the only thing in the header that needs png.h.

Hysrok commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+{
+  InputParameters params = validParams<Output>();
+  params.addParam<Real>("resolution", 2000, "The resolution of the image.");
+  params.addParam<std::string>("PNGFile", "Adam", "Root filename of the PNG to be created.");
```
So what I found on _file_base, as a string variable, is that it is created in the fileOutput class, unless you meant the input parameter "file_base", but I'm pretty sure that's not the one are talking about.  Should I change this class to inherit from that class, or include it in the includes, or did you just want me to see how it's used and follow that?

permcody replied:
Yes, change the inheritance. You should be inheriting from FileOutput.

Hysrok commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,228 @@
+
+void PNGOutput::makePNG() {
+  // Increment the testStep
+  testStep++;
```
Would it be the _t_step from Output?

permcody commented:
@Hysrok - I'm going to recommend that you start your fixes by looking at the non-unity build first. You are missing a lot of headers still...

Rule of thumb, if you use a type in your code, include the header (Prefer the .C file unless you have to put it in the .h file).

If you want to compile a non-unity build locally set MOOSE_UNITY=false in your environment and rebuild everything.

permcody commented:
@Hysrok - I added one more line to my existing Makefile branch for the include path to the header. We'll need that to make sure we pick up the right version of the `png.h` file. I think we are getting lucky and picking it up through conda or vtk (it exists in both places).

dschwen commented:
Just a thought, instead of reinventing the wheel on color output why not use a library like
https://github.com/richardroberts1992/Spectrum
to translate float values into colors taken from a colormap? This might very well be a follow on PR though.

friedmud commented:
Spectrum does look cool - but why does in include QDebug?

On Fri, Jun 28, 2019 at 5:13 PM Daniel Schwen <notifications@github.com>
wrote:

> Just a thought, instead of reinventing the wheel on color output why not
> use a library like
> https://github.com/richardroberts1992/Spectrum
> to translate float values into colors taken from a colormap? This might
> very well be a follow on PR though.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13646?email_source=notifications&email_token=AAGUSMMNY4P73K2A5X4TUALP42LJTA5CNFSM4H3NEVWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY3LPIY#issuecomment-506902435>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGUSMIPLMVCVE6NZ542LBTP42LJTANCNFSM4H3NEVWA>
> .
>
-- 
Sent from my iPhone

dschwen commented:
Yeah, weird. Doesn't seem to use it though. Make a PR?

Commit:
Applied changes suggested from pull request comments.

-Using inheriting from FileOutput
-Input files for the png's aren't worrying about exodus etc
-Changed several types
-Removed unnecessary hardcoded lines from MakeFile
-Fixed naming convention of variables
-Added an #IFDEF to the header file

Commit:
Applied additional changed from pull request comments.

- Added the header files that were missing.
- Removed the macros.
- Redid filename configurations for created png's.
- Using MooseUtils to check file writability.

Commit:
Added tests to make sure png's are being printed (files with .png exist).

Commit:
Some clean up and additional changes from PR comments.

- Added vector to store the row data to help avoid raw memory 
management.
- Revmoved the parameter PNGFile and anywhere it was referenced to.
- Added possible blue->red color scheme. Commented out for now.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/362385/) on b1d1ffc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

Hysrok commented:
```ssh: connect to host github.com port 22: Operation timed out
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.```
I've been getting this problem a lot today.  I found a work around that worked for me, but now it looks like even the moose build checker is having the problem.  Is something wrong with GitHub, or am I just doing something wrong?

Commit:
- Added additional color schemes - Added MooseEnum as the new way to select a color scheme - Added additional comments to explain the code

permcody commented:
If you haven't heard from the MOOSE team. The problem is INL related, not
you. They are still working on a fix as of now. You can use HTTPS instead
of SSH to get around the issue.

On Wed, Jul 3, 2019 at 7:05 PM Herald of Heraldo <notifications@github.com>
wrote:

> fatal: Could not read from remote repository.
>
> Please make sure you have the correct access rights
> and the repository exists.```
> I've been getting this problem a lot today.  I found a work around that worked for me, but now it looks like even the moose build checker is having the problem.  Is something wrong with GitHub, or am I just doing something wrong?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13646?email_source=notifications&email_token=AAXFOICC5EF5J6L35NI5XKDP5VEEVA5CNFSM4H3NEVWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZGBB6A#issuecomment-508301560>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXFOICA2DJ43F2D7MEFRZTP5VEEVANCNFSM4H3NEVWA>
> .
>

permcody commented on framework/src/outputs/PNGOutput.C:
```diff
@@ -0,0 +1,315 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
```
You'll need to wrap this whole file in `MOOSE_HAVE_LIBPNG` too.

permcody commented:
@Hysrok - Looks like this is failing to build on Linux still. From a quick glance it might be that we are picking up inconsistent headers from libPNG or something. To troubleshoot this, you are going to want to login to a Linux system with the exact same environment as the build systems.

1) ssh to "rod.hpc.inl.gov"
2) run "moosebuild"
3) clone, build libmesh, build moose and start troubleshooting

I can help when you get to step #3. I suspect we have some more work todo to get the include and library flags rock solid.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/365995/) on 6e158c4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/13646/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/13646/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format cfcf841ec59b526ac8b1c538a1cb1b50a8811b22`

Commit:
Applied changes.

-Wrapped PNGOoutput.C in MOOSE_HAVE_LIBPNG wrapper.
-Updated some colors.
-Added tranparency options.

permcody commented:
Ought oh.... Somehow you have a mess of other peoples commits in here now. Did you forget to rebase?

permcody commented:
Worse yet you clobbered your change you made earlier just before lunch.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/366205/) on cb841c0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13646/site)

This comment will be updated on new commits.

Commit:
Made changes to fix fails in the moosebuild

-Added PNGOutput.md file
-Added "design" parameter in test input files
-Set max_parallel parameter to 1 for adv_diff_reaction
-Lowered the resolution of created images to avoid getting timed out

permcody committed:
Updating requirements a bit

Removing duplicates and adding detail.

permcody reviewed (dismissed):
Looks good! I updated your requirements a bit since we don't accept duplicate requirements anymore. Once this tests successfully I'll get it merged. Let me know when you get in and we can talk about what's next and your poster.

dschwen commented on framework/doc/content/source/outputs/png/PNGOutput.md:
```diff
@@ -0,0 +1,18 @@
+  [png]
+    type = PNGOutput
+    transparent = true    #indicates whether the background will be transparent
+    resolution = 50       #resolution of each point
```
Just looking at the documentation strings I have no idea what this number means. The code suggests it is  a scale factor to get from MOOSE length units to pixels. I suggest to make this clear in the docs.

permcody replied:
The resolution might be "pixels", except that's not even a great definition with things like retina displays and everything else that complicates rendering sizes. This is the number of "samples" taken in the x and y directions right now but that's not the best solution for resolution, which should be in units per something. Also we need to account for different spacing in x and y. More work to do.

permcody committed:
Fixing Adv_Diff_reaction Test spec

permcody commented:
@dschwen - We are going to be working on the next phase of this project now. This capability isn't useful yet on anything but small images because of the extensive use of MeshFunction. It can take minutes to generate a single image on a workstation size run right now. I'm going to merge this in and we will continue to work on documentation and capability as we make this scalable.

friedmud commented:
Hmm - I really think pixels is what we want for resolution... ultimately I
want to control the exact size of the PNG.

On Tue, Jul 16, 2019 at 1:04 PM Cody Permann <notifications@github.com>
wrote:

> *@permcody* commented on this pull request.
> ------------------------------
>
> In framework/doc/content/source/outputs/png/PNGOutput.md
> <https://github.com/idaholab/moose/pull/13646#discussion_r304072687>:
>
> > @@ -0,0 +1,18 @@
> +# PNGOutput (Outputs)
> +
> +The PNGOutput object is the class for writing png files from the input files.
> +
> +Following is an example of all the different input variables that can be used.
> +
> +```
> +[Outputs]
> +  [png]
> +    type = PNGOutput
> +    transparent = true    #indicates whether the background will be transparent
> +    resolution = 50       #resolution of each point
>
> The resolution might be "pixels", except that's not even a great
> definition with things like retina displays and everything else that
> complicates rendering sizes. This is the number of "samples" taken in the x
> and y directions right now but that's not the best solution for resolution,
> which should be in units per something. Also we need to account for
> different spacing in x and y. More work to do.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13646?email_source=notifications&email_token=AAGUSMODS2QS3XYTDOI6WETP7YLTXA5CNFSM4H3NEVWKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6TU4OQ#discussion_r304072687>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGUSMP64GSPGWMF4U37VMDP7YLTXANCNFSM4H3NEVWA>
> .
>
